controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -205,109 +193,167 @@ public void refresh() {
     @Override
     public CompletableFuture<Void> checkScopeExists() {
         return store.checkExists(scopePath)
-                .thenAccept(x -> {
-                    if (!x) {
-                        throw StoreException.create(StoreException.Type.DATA_NOT_FOUND, scopePath);
-                    }
-                });
+                    .thenAccept(x -> {
+                        if (!x) {
+                            throw StoreException.create(StoreException.Type.DATA_NOT_FOUND, scopePath);
+                        }
+                    });
     }
 
     @Override
-    CompletableFuture<Void> createSealedSegmentsRecord(byte[] sealedSegmentsRecord) {
-        return Futures.toVoid(store.createZNodeIfNotExist(sealedSegmentsPath, sealedSegmentsRecord));
+    CompletableFuture<Void> createRetentionSetDataIfAbsent(byte[] data) {
+        return Futures.toVoid(store.createZNodeIfNotExist(retentionSetPath, data));
     }
 
     @Override
-    CompletableFuture<Data> getSealedSegmentsRecord() {
-        return store.getData(sealedSegmentsPath);
+    CompletableFuture<Data> getRetentionSetData() {
+        return store.getData(retentionSetPath);
     }
 
     @Override
-    CompletableFuture<Void> updateSealedSegmentsRecord(Data update) {
-        return store.setData(sealedSegmentsPath, update)
-                .thenApply(x -> cache.invalidateCache(sealedSegmentsPath));
+    CompletableFuture<Version> updateRetentionSetData(Data retention) {
+        return store.setData(retentionSetPath, retention)
```

Comment:
Actually it is callers responsibility to decide when conditional and unconditional updates are required.
The choice is business logic layers responsibility. The underlying raw storage layer doesnt take that decision but exposes ability to perform either type of update. 

Though, incidentally, we no longer have any "unconditional" updates anymore. But the design choice is deliberate to have the business logic take that decision.

Reply:
I think this is error prone. You know this, but someone picking up this code might miss it altogether because of the level of indirection. If you want to keep it this way, then we should make sure to document it well on javadocs.

In any case, this is independent of this issue. Can we record it somewhere so that we don't lose it?

Reply:
https://github.com/pravega/pravega/issues/3038